plugins/network-elements/nuage-vsp/src/com/cloud/network/manager/NuageVspManagerImpl.java

```diff
@@ -152,53 +226,181 @@ public NuageVspDeviceVO addNuageVspDevice(AddNuageVspDeviceCmd cmd) {
+            paramsTobeUpdated.put("cmsuser", command.getUserName());
+        }
+
+        if (StringUtils.isNotBlank(command.getPassword())) {
+            String encodedNewPassword = org.apache.commons.codec.binary.StringUtils.newStringUtf8(Base64.encodeBase64(command.getPassword().getBytes()));
+            if (!encodedNewPassword.equals(nuageVspHost.getDetails().get("cmsuserpass"))) {
+                paramsTobeUpdated.put("cmsuserpass", org.apache.commons.codec.binary.StringUtils.newStringUtf8(Base64.encodeBase64(command.getPassword().getBytes())));
+            }
+        }
+
+        if (command.getPort() != null &&
+                command.getPort() != Integer.parseInt(nuageVspHost.getDetails().get("port"))) {
+            paramsTobeUpdated.put("port", String.valueOf(command.getPort()));
+        }
+
+        GetClientDefaultsCommand getClientDefaultsCmd = new GetClientDefaultsCommand();
+        GetClientDefaultsAnswer getClientDefaultsAnswer = (GetClientDefaultsAnswer) _agentMgr.easySend(nuageVspHost.getId(), getClientDefaultsCmd);
+        String apiVersion = Objects.firstNonNull(command.getApiVersion(), getClientDefaultsAnswer.getCurrentApiVersion());
+        SupportedApiVersionCommand supportedApiVersionCmd = new SupportedApiVersionCommand(apiVersion);
+        SupportedApiVersionAnswer supportedApiVersionAnswer = (SupportedApiVersionAnswer) _agentMgr.easySend(nuageVspHost.getId(), supportedApiVersionCmd);
+        if (!supportedApiVersionAnswer.getResult()) {
+            throw new CloudRuntimeException("Incorrect API version: Nuage plugin only supports " + getClientDefaultsAnswer.getCurrentApiVersion());
+        }
+
+        String apiRelativePath = "/nuage/api/" + command.getApiVersion();
+        if (!apiRelativePath.equals(nuageVspHost.getDetails().get("apirelativepath"))) {
+            paramsTobeUpdated.put("apirelativepath", apiRelativePath);
+            paramsTobeUpdated.put("apiversion", apiVersion);
+        }
+
+        if (command.getApiRetryCount() != null &&
+                command.getApiRetryCount() != Integer.parseInt(nuageVspHost.getDetails().get("retrycount"))) {
+            paramsTobeUpdated.put("retrycount", String.valueOf(command.getApiRetryCount()));
+        }
+
+        if (command.getApiRetryInterval() != null &&
+                command.getApiRetryInterval() != Integer.parseInt(nuageVspHost.getDetails().get("retryinterval"))) {
+            paramsTobeUpdated.put("retryinterval", String.valueOf(command.getApiRetryInterval()));
+        }
+
+        if (paramsTobeUpdated.size() > 0) {
+            Map<String, String> latestParamsValue = new HashMap<String, String>();
+            latestParamsValue.putAll(nuageVspHost.getDetails());
+            latestParamsValue.putAll(paramsTobeUpdated);
+            Map<String, Object> hostdetails = new HashMap<String, Object>();
+            hostdetails.putAll(latestParamsValue);
```

Comment:
``HashMap`` provides a copy constructor.  Therefore, pass ``latestParamsValue`` into the constructor and remove the ``putAll`` call.